packages/react-reconciler/src/ReactFiberUnwindWork.js

```diff
@@ -121,19 +158,139 @@ export default function<C, CX>(
     return update;
   }
 
+  function waitForPromiseAndScheduleRecovery(finishedWork, thenable) {
+    // Await promise
+    thenable.then(() => {
+      // Once the promise resolves, we should try rendering the non-
+      // placeholder state again.
+      const startTime = recalculateCurrentTime();
+      const expirationTime = computeExpirationForFiber(startTime, finishedWork);
+      const recoveryUpdate = createUpdate(expirationTime);
+      enqueueUpdate(finishedWork, recoveryUpdate, expirationTime);
+      scheduleWork(finishedWork, startTime, expirationTime);
+    });
+  }
+
   function throwException(
+    root: FiberRoot,
     returnFiber: Fiber,
     sourceFiber: Fiber,
-    rawValue: mixed,
+    value: mixed,
+    renderIsExpired: boolean,
+    remainingTimeMs: number,
+    elapsedMs: number,
+    renderStartTime: number,
     renderExpirationTime: ExpirationTime,
   ) {
     // The source fiber did not complete.
     sourceFiber.effectTag |= Incomplete;
     // Its effect list is no longer valid.
     sourceFiber.firstEffect = sourceFiber.lastEffect = null;
 
-    const value = createCapturedValue(rawValue, sourceFiber);
+    if (
+      value !== null &&
+      typeof value === 'object' &&
+      typeof value.then === 'function'
+    ) {
+      // This is a thenable.
+      // Allow var because this is used in a closure.
+      // eslint-disable-next-line no-var
+      var thenable: {then: (() => mixed) => mixed} = (value: any);
+
+      // Find the earliest timeout of all the timeouts in the ancestor path.
+      // TODO: Alternatively, we could store the earliest timeout on the context
+      // stack, rather than searching on every suspend.
+      let workInProgress = returnFiber;
+      let earliestTimeoutMs = -1;
+      searchForEarliestTimeout: do {
+        switch (workInProgress.tag) {
+          case TimeoutComponent: {
+            const current = workInProgress.alternate;
+            if (current !== null && current.memoizedState === true) {
+              // A parent Timeout already committed in a placeholder state. We
+              // need to handle this promise immediately. In other words, we
+              // should never suspend inside a tree that already expired.
+              earliestTimeoutMs = 0;
+              break searchForEarliestTimeout;
+            }
+            let timeoutPropMs = workInProgress.pendingProps.ms;
```

Comment:
Conceptually I want the incoming props. In this particular case, both memoizedProps and pendingProps are the same thing so it doesn't matter.

> It used to be null. Maybe it will be again

Ok but that's true of all these decisions. I could use the memoized props, but what if we move memoization to the complete phase? We used to memoize in the complete phase, maybe we will again.

Reply:
The other option is to store the earliest timeout on the context stack. I think that might be best.